grails-app/controllers/org/pih/warehouse/core/DocumentController.groovy

```diff
@@ -222,8 +224,8 @@ class DocumentController {
 
             documentInstance.validate()
 
-            List<DocumentType> nonTemplateDocumentTypes = documentService.getNonTemplateDocumentTypes()
```

Comment:
@EWaterman it is used in many controllers (invoice, stock movement, etc.): https://github.com/openboxes/openboxes/pull/4750/files

Reply:
I think the confusion for me is around having both the DocumentCode enum and the DocumentType domain. It's not immediately clear when each is meant to be used.

My assumption was:
- DocumentCode enumerates all the types of documents that we have the capacity to support in open boxes
- DocumentType is all the types of documents that we're actually using on some environment.

But I must be wrong because we have some DocumentTypes that don't have a code (why?).

For reference, here's all the DocumentTypes that we have in prod currently:

```
+----------------------------------+---------+---------------------+-------------------------------------------------------------------------------------------------------------+---------------------+-------------------------------------------------+------------+-------------------------+
| id                               | version | date_created        | description                                                                                                 | last_updated        | name                                            | sort_order | document_code           |
+----------------------------------+---------+---------------------+-------------------------------------------------------------------------------------------------------------+---------------------+-------------------------------------------------+------------+-------------------------+
| 09f5e116789f66c30178c3f594e4482b |       1 | 2021-04-12 02:42:00 | Approval for purchase                                                                                       | 2021-04-12 02:42:54 | Approval                                        |          0 | NULL                    |
| 09f5e116789f66c30178c3f6d64c482c |       0 | 2021-04-12 02:43:57 | NULL                                                                                                        | 2021-04-12 02:43:57 | Shipping Documents Folder                       |          0 | NULL                    |
| 09f5e1167b0c7ce4017b11e113d427b3 |       0 | 2021-08-04 15:56:09 | NULL                                                                                                        | 2021-08-04 15:56:09 | PO Template                                     |          0 | PURCHASE_ORDER_TEMPLATE |
| 09f5e1167b3ab4e5017b41ac474a383a |       0 | 2021-08-13 22:40:15 | NULL                                                                                                        | 2021-08-13 22:40:15 | Donation Valuation                              |          0 | SHIPPING_DOCUMENT       |
| 09f5e116845fa23f01847281eb943e19 |       0 | 2022-11-13 19:40:28 | Link to a folder with all purchasing documents for this order                                               | 2022-11-13 19:40:28 | PO Folder                                       |          0 | SHIPPING_DOCUMENT       |
| 2                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Bill of Lading|fr:Facture de la cargaison       |          0 | NULL                    |
| 3                                |       1 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2021-04-12 02:41:23 | Packing List|fr:Liste de chargement             |          0 | NULL                    |
| 4                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Certificate of Donation|fr:Certificat de don    |          0 | NULL                    |
| 5                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Commercial Invoice|fr:Facture Commerciale       |          0 | NULL                    |
| 6                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Material Safey Data Sheet|fr:Fiche signal       |          0 | NULL                    |
| 66762f6c61e34cfd9297ecb0fcee2df2 |       0 | 2022-06-23 03:32:07 | NULL                                                                                                        | 2022-06-23 03:32:07 | Invoice Template                                |          0 | INVOICE_TEMPLATE        |
| 7                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Certificate of Analysis|fr:Certificat d'analyse |          0 | NULL                    |
| 8                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Manifest|fr:Manifeste                           |          0 | NULL                    |
| 9                                |       0 | 2010-08-25 00:00:00 | NULL                                                                                                        | 2010-08-25 00:00:00 | Other|fr:Autre                                  |          0 | NULL                    |
| cab2b1078b815bed018b825427030ab3 |       0 | 2023-10-30 20:43:55 | This is used to upload Lesotho Purchase Requisitions as downloaded from the powerapp.                       | 2023-10-30 20:43:55 | Purchase Requisition                            |          0 | NULL                    |
| cab2b1078b815bed018b9f90951b43cc |       0 | 2023-11-05 12:58:54 | This is used to upload Lesotho signed receipts when shipment is receive directly by the procurement office. | 2023-11-05 12:58:54 | Receipt Signed                                  |          0 | NULL                    |
| ff8080818662c5070186c8082b3f0ef3 |       0 | 2023-03-09 20:20:24 | Quotes for Purchase Order                                                                                   | 2023-03-09 20:20:24 | Quotes                                          |          0 | NULL                    |
| ff8080818662c5070186c8085f500ef6 |       0 | 2023-03-09 20:20:37 | NULL                                                                                                        | 2023-03-09 20:20:37 | Ordering Template                               |          0 | NULL                    |
| ff8080818662c5070186c80885720ef8 |       0 | 2023-03-09 20:20:47 | NULL                                                                                                        | 2023-03-09 20:20:47 | Order Justification Template                    |          0 | NULL                    |
| ff8080818662c5070186c808a5a20efc |       0 | 2023-03-09 20:20:55 | NULL                                                                                                        | 2023-03-09 20:20:55 | Analysis Template                               |          0 | NULL                    |
| ff8080818662c5070186c808d1400f00 |       0 | 2023-03-09 20:21:06 | NULL                                                                                                        | 2023-03-09 20:21:06 | PO Copy                                         |          0 | NULL                    |
| ff80818166ac5f350166ac84c2150001 |       0 | 2018-10-25 00:00:00 | NULL                                                                                                        | 2018-10-25 00:00:00 | Shipping Template                               |          0 | SHIPPING_TEMPLATE       |
| REQUISITION_TEMPLATE             |       0 | 2023-06-05 04:10:00 | NULL                                                                                                        | 2023-06-05 04:10:00 | Requisition Template                            |          0 | REQUISITION_TEMPLATE    |
+----------------------------------+---------+---------------------+-------------------------------------------------------------------------------------------------------------+---------------------+-------------------------------------------------+------------+-------------------------+
```

Reply:
@EWaterman I agree with you it might be confusing, and I also can't answer why that is the case. I agree the `document_code` should probably be a required field - in the same way like we have the `EventType` and `EventTypeCode` (eventTypeCode is a required field), but maybe we lack some domain knowledge to answer that, and maybe there is a use-case for the `documentCode` being `null`.

Reply:
If something is confusing, please ask. There's likely a method to the madness. In fact, these types of questions would be great topics for a tech huddle. We could then document the discussion and contemplate whether there's actual tech debt to unravel. 

The DocumentCode enum (which probably should have been named DocumentTypeCode) is not required because it's only relevant for templates and some specific document types. More specifically, the document code was added to allow us to attribute importance/meaning to certain document types, particularly document templates. We use an enum rather than the document type because it's safer and easier to map an enum to functionality. Using the document type ID or name was another option, but what happens if those fields change or the document type is deleted. 

Going further, document type itself is not required because we didn't want to put an undue burden on the user to create an appropriate document type for every document they attach to an object. Remember documents are global. They can be attached to products, shipments, POs, etc. Or they can just be uploaded to the system globally. Sometimes you want to upload photos to a product, sometimes you want to dump files onto a PO or shipment with the "Other" document type. And other times you want to upload a logo that you can reference in your system configuration.